### PR TITLE
More refactor of the topo package.

### PR DESCRIPTION
**Pull request by alainjobart**
A few changes:

* Moving the Lock function down to Backend, so it can be shared between keyspace and shard. Refactoring the Lock method in each implementation.
* GetKnownCells is now implemented by topo.Server.

Next steps are to move the cell map into topo.Server.

### CODE REVIEW

File: go/vt/vtctld/explorer_test.go [hunk 1]
```diff
@@ -212,7 +209,8 @@ func TestHandleBadPath(t *testing.T) {
 	want := "node doesn't exist"
 
 	ts := memorytopo.New(cells...)
-	ex := newBackendExplorer(ts)
+	tts := topo.Server{Impl: ts}
```
Comment (pivanof) [inline go/vt/vtctld/explorer_test.go]: Why not memorytopo.NewServer()?

File: go/vt/vttablet/tabletmanager/init_tablet.go [hunk 2]
```diff
@@ -155,7 +155,7 @@ func (agent *ActionAgent) InitTablet(port, gRPCPort int32) error {
 		Hostname:       hostname,
 		PortMap:        make(map[string]int32),
 		Keyspace:       *initKeyspace,
-		Shard:          *initShard,
+		Shard:          shard,
```
Comment (pivanof) [inline go/vt/vttablet/tabletmanager/init_tablet.go]: How is this related?

File: go/vt/vtgate/vtgateservice/vtgateservice_testing/mock_vtgateservice.go [hunk 3]
```diff
@@ -1,25 +1,12 @@
...
-
-Unless required by applicable law or agreedto in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
-
-// Automatically generated by MockGen. DO NOT EDIT!
+// Code generated by MockGen. DO NOT EDIT.
```
Comment (pivanof) [inline go/vt/vtgate/vtgateservice/vtgateservice_testing/mock_vtgateservice.go]: All these mock regenerations are not related to locking, right? Why are they here?

File: go/vt/topo/consultopo/lock.go [hunk 4]
```diff
@@ -20,20 +20,41 @@ import (
...
+		return nil, err
+	}
+	lockPath := path.Join(c.root, dirPath, locksFilename)
+
 	// Build the lock structure.
-	l, err := s.global.client.LockOpts(&api.LockOptions{
+	l, err := c.client.LockOpts(&api.LockOptions{
 		Key:   lockPath,
-		Value: []byte(contents),
+		Value: []byte("lock"),
```
Comment (pivanof) [inline go/vt/topo/consultopo/lock.go]: Why did you abandon the idea of writing contents into the lock?

File: go/vt/topo/backend.go [hunk 5]
```diff
@@ -167,9 +179,10 @@ type Version interface {
 }
 
 // LockDescriptor is an interface that describes a lock.
+// It will be returned by Lock(). It contains the Unlock method.
 type LockDescriptor interface {
-	// String returns a text representation of the lock.
-	String() string
+	// Unlock frees the lock.
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/backend.go]: Done.

File: go/vt/topo/consultopo/lock.go [hunk 6]
```diff
@@ -20,20 +20,41 @@ import (
...
+}
+
+// Lock is part of the topo.Backend interface.
+func (s *Server) Lock(ctx context.Context, cell string, dirPath string) (topo.LockDescriptor, error) {
+	// We list the directory first to make sure it exists.
+	if _, err := s.ListDir(ctx, cell, dirPath); err != nil {
+		if err == topo.ErrNoNode {
+			return nil, err
+		}
+		return nil, fmt.Errorf("cannot ListDir(%v,%v) before locking", cell, dirPath)
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/consultopo/lock.go]: Done.

File: go/vt/topo/consultopo/lock.go [hunk 7]
```diff
@@ -73,21 +94,24 @@ func (s *Server) lock(ctx context.Context, c *cellClient, lockPath, contents str
...
-	// Sanity check.
-	if lockPath != actionPath {
-		return fmt.Errorf("unlock: actionPath doesn't match directory being unlocked: %q != %q", lockPath, actionPath)
-	}
+// Unlock is part of the topo.LockDescriptor interface.
+func (ld *consulLockDescriptor) Unlock(ctx context.Context) error {
+	return ld.s.unlock(ctx, ld.c, ld.lockPath)
+}
 
+// unlock releases a lock acquired by lock() on the given directory.
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/consultopo/lock.go]: Done.

File: go/vt/topo/etcd2topo/lock.go [hunk 8]
```diff
@@ -109,101 +108,96 @@ func (s *Server) waitOnLastRev(ctx context.Context, nodePath string, revision in
 	return false, nil
 }
 
-func (s *Server) lock(ctx context.Context, nodePath string, contents string) (string, error) {
+// ectdLockDescriptor implements topo.LockDescriptor.
+type ectdLockDescriptor struct {
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/etcd2topo/lock.go]: Done.

File: go/vt/topo/etcd2topo/lock.go [hunk 9]
```diff
@@ -109,101 +108,96 @@ func (s *Server) waitOnLastRev(ctx context.Context, nodePath string, revision in
...
+		if err == topo.ErrNoNode {
+			return nil, err
+		}
+		return nil, fmt.Errorf("cannot ListDir(%v,%v) before locking", cell, dirPath)
+	}
+
+	return s.lock(ctx, cell, dirPath, "lock")
+}
+
+// lock is used by both Lock()  and master election.
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/etcd2topo/lock.go]: Done.

File: go/vt/topo/helpers/tee.go [hunk 10]
```diff
@@ -154,114 +143,56 @@ func (tee *Tee) Watch(ctx context.Context, cell, filePath string) (*topo.WatchDa
...
 
-	// lock lockSecond
-	sLockPath, err := tee.lockSecond.LockShardForAction(ctx, keyspace, shard, contents)
+	// Lock lockSecond.
+	sLD, err := tee.lockSecond.Lock(ctx, cell, dirPath)
 	if err != nil {
-		if err := tee.lockFirst.UnlockShardForAction(ctx, keyspace, shard, pLockPath, "{}"); err != nil {
-			log.Warningf("Failed to unlock lockFirst shard after failed lockSecond lock for %v/%v", keyspace, shard)
+		if err := fLD.Unlock(ctx); err != nil {
+			log.Warningf("Failed to unlock lockFirst after failed lockSecond lock for %v %v", cell, dirPath)
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/helpers/tee.go]: Done.

File: go/vt/topo/helpers/tee.go [hunk 11]
```diff
@@ -154,114 +143,56 @@ func (tee *Tee) Watch(ctx context.Context, cell, filePath string) (*topo.WatchDa
...
+func (ld *teeTopoLockDescriptor) Unlock(ctx context.Context) error {
+	// Unlock lockSecond, then lockFirst.
+	serr := ld.secondLockDescriptor.Unlock(ctx)
+	ferr := ld.firstLockDescriptor.Unlock(ctx)
 
 	if serr != nil {
-		if perr != nil {
-			log.Warningf("Secondary UnlockShardForAction(%v/%v, %v) failed: %v", keyspace, shard, sLockPath, serr)
+		if ferr != nil {
+			log.Warningf("First Unlock(%v) failed: %v", ld.dirPath, ferr)
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/helpers/tee.go]: Done.

File: go/vt/topo/memorytopo/lock.go [hunk 12]
```diff
@@ -56,27 +67,33 @@ func (mt *MemoryTopo) lock(ctx context.Context, nodePath string, contents string
...
-func (mt *MemoryTopo) unlock(ctx context.Context, nodePath, actionPath string) error {
-	if nodePath != actionPath {
-		return fmt.Errorf("invalid actionPath %v was expecting %v", actionPath, nodePath)
-	}
+// Unlock is part of the topo.LockDescriptor interface.
+func (ld *memoryTopoLockDescriptor) Unlock(ctx context.Context) error {
+	return ld.mt.unlock(ctx, ld.cell, ld.nodePath)
+}
 
+// unlock is used by Unlock().
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/memorytopo/lock.go]: Done.

File: go/vt/topo/test/lock.go [hunk 13]
```diff
@@ -69,50 +70,52 @@ func checkKeyspaceLockTimeout(ctx context.Context, t *testing.T, ts topo.Impl) {
...
+		t.Fatalf("Unlock(again) worked")
 	}
 }
 
-// checkKeyspaceLockMissing makes sure we can't lock a non-existing keyspace
-func checkKeyspaceLockMissing(ctx context.Context, t *testing.T, ts topo.Impl) {
-	if _, err := ts.LockKeyspaceForAction(ctx, "test_keyspace_666", "fake-content"); err == nil {
-		t.Fatalf("LockKeyspaceForAction(test_keyspace_666) worked for non-existing keyspace")
+// checkLockMissing makes sure we can't lock a non-existing directory.
+func checkLockMissing(ctx context.Context, t *testing.T, ts topo.Impl) {
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/test/lock.go]: Because the Lock() interface is on the Impl/Backend, not the ts, that has LockKeyspace.

File: go/vt/topo/zk2topo/lock.go [hunk 14]
```diff
@@ -29,25 +29,31 @@ import (
 
 // This file contains the lock management code for zktopo.Server.
 
-// lockForAction creates the locks node in zookeeper, waits for the
-// queue lock, displays a nice error message if it cant get it.
-func (zs *Server) lockForAction(ctx context.Context, locksDir, contents string) (string, error) {
-	conn, root, err := zs.connForCell(ctx, topo.GlobalCell)
+// zsLockDescriptor implements topo.LockDescriptor.
+type zsLockDescriptor struct {
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/zk2topo/lock.go]: Done.

File: go/vt/topo/zk2topo/lock.go [hunk 15]
```diff
@@ -29,25 +29,31 @@ import (
...
 	}
 
 	// Lock paths end in a trailing slash to that when we create
 	// sequential nodes, they are created as children, not siblings.
-	locksDir = path.Join(root, locksDir) + "/"
+	locksDir := path.Join(root, dirPath, locksPath) + "/"
 
 	// Create the locks path, possibly creating the parent.
-	locksPath, err := CreateRecursive(ctx, conn, locksDir, []byte(contents), zk.FlagSequence|zk.FlagEphemeral, zk.WorldACL(PermFile), 1)
+	lockPath, err := CreateRecursive(ctx, conn, locksDir, []byte("lock"), zk.FlagSequence|zk.FlagEphemeral, zk.WorldACL(PermFile), 1)
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/zk2topo/lock.go]: Done.

File: go/vt/vtctld/explorer_test.go (hunk 1 above)
Comment (alainjobart) [reply go/vt/vtctld/explorer_test.go]: Done.

File: go/vt/vtgate/vtgateservice/vtgateservice_testing/mock_vtgateservice.go (hunk 3 above)
Comment (alainjobart) [reply go/vt/vtgate/vtgateservice/vtgateservice_testing/mock_vtgateservice.go]: Because I needed to regenerate the mocks, and they all come as one. I used a different commit for the others, so it's clearer.

File: go/vt/vttablet/tabletmanager/init_tablet.go (hunk 2 above)
Comment (alainjobart) [reply go/vt/vttablet/tabletmanager/init_tablet.go]: Small topo fix from the last set of commits.

File: go/vt/topo/consultopo/lock.go (hunk 4 above)
Comment (alainjobart) [reply go/vt/topo/consultopo/lock.go]: I re-added it. I thought it would be simpler, but the contents is useful in a lot of topo implementations.

File: go/vt/topo/locks.go [hunk 16]
```diff
@@ -45,7 +45,6 @@ var (
 )
 
 // Lock describes a long-running lock on a keyspace or a shard.
-// It needs to be public as we JSON-serialize it.
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/locks.go]: Using it to generate the contents again, so re-added the comment.

File: go/vt/topo/helpers/tee.go [hunk 17]
```diff
@@ -188,7 +190,7 @@ func (ld *teeTopoLockDescriptor) Unlock(ctx context.Context) error {
 
 	if serr != nil {
 		if ferr != nil {
-			log.Warningf("First Unlock(%v) failed: %v", ld.dirPath, ferr)
+			log.Warningf("First Unlock(%v, %v) failed: %v", ld.cell, ld.dirPath, ferr)
```
Comment (pivanof) [inline go/vt/topo/helpers/tee.go]: Note that you've got inconsistency in error messages: in some places cell and path are separated just by space, in some they are separated by comma. I'd personally prefer the latter format (with the comma), but in any case it should be consistent across all the messages.

File: go/vt/topo/test/lock.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pivanof) [reply go/vt/topo/test/lock.go]: But checlLockTimeout() above also verifies behavior of Lock() function and it accepts topo.Server. They verify the same Lock() function, right?

File: go/vt/topo/etcd2topo/lock.go [hunk 18]
```diff
@@ -109,101 +108,96 @@ func (s *Server) waitOnLastRev(ctx context.Context, nodePath string, revision in
 	return false, nil
 }
 
-func (s *Server) lock(ctx context.Context, nodePath string, contents string) (string, error) {
+// ectdLockDescriptor implements topo.LockDescriptor.
+type ectdLockDescriptor struct {
+	s       *Server
+	cell    string
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/etcd2topo/lock.go]: I also fixed this, forgot to comment.

File: go/vt/topo/etcd2topo/lock.go [hunk 19]
```diff
@@ -109,101 +108,81 @@ func (s *Server) waitOnLastRev(ctx context.Context, nodePath string, revision in
...
+		if err == topo.ErrNoNode {
+			return nil, err
+		}
+		return nil, fmt.Errorf("cannot ListDir(%v,%v) before locking", cell, dirPath)
+	}
+
+	return s.lock(ctx, cell, dirPath, contents)
+}
+
+// lock is used by both Lock() and master election.
```
Comment (pivanof) [inline go/vt/topo/etcd2topo/lock.go]: Mind removing this non-useful comment as well?

File: go/vt/vtgate/vtgateservice/vtgateservice_testing/mock_vtgateservice.go (hunk 3 above)
Comment (pivanof) [reply go/vt/vtgate/vtgateservice/vtgateservice_testing/mock_vtgateservice.go]: So you can regenerate the mocks with all the changes unrelated to this PR as a separate PR, merge them, and then this PR will contain only changes in the mocks that are related to it, right?

File: go/vt/vttablet/tabletmanager/init_tablet.go (hunk 2 above)
Comment (pivanof) [reply go/vt/vttablet/tabletmanager/init_tablet.go]: I don't understand. You are changing what's stored in the topology, which means there should be a reason for that (which is not explained here) and a test verifying that you store the correct thing, right? I don't understand to which set of commits this is related, and together with the test I think it should be extracted into a separate PR.

**Comment by sougou** [comment]
LGTM
I assume pavel's comments are addressed?
You also need to resolve the merge conflict

File: go/vt/vttablet/tabletmanager/init_tablet.go (hunk 2 above)
Comment (alainjobart) [reply go/vt/vttablet/tabletmanager/init_tablet.go]: I pulled this change out into:
https://github.com/youtube/vitess/pull/3417
Will merge and resolve.

File: go/vt/topo/etcd2topo/lock.go (hunk 19 above)
Comment (alainjobart) [reply go/vt/topo/etcd2topo/lock.go]: This one is useful, it justifies why this is just not directly inside the Lock method.

File: go/vt/topo/test/lock.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/test/lock.go]: Ah I see, fixed it with changing the signature of the other method. This will be changed again when I move the CellMap, most likely.

**Comment by alainjobart** [comment]
OK this is ready for final review now. Re-based to master, fixed conflicts. I think I addressed all comments (which is somewhat hard to be sure of, since this review tool is not that good).

@pivanof PTAL.

File: go/vt/topo/helpers/tee.go [hunk 20]
```diff
@@ -154,114 +143,58 @@ func (tee *Tee) Watch(ctx context.Context, cell, filePath string) (*topo.WatchDa
...
 
-	// lock lockSecond
-	sLockPath, err := tee.lockSecond.LockShardForAction(ctx, keyspace, shard, contents)
+	// Lock lockSecond.
+	sLD, err := tee.lockSecond.Lock(ctx, cell, dirPath, contents)
 	if err != nil {
-		if err := tee.lockFirst.UnlockShardForAction(ctx, keyspace, shard, pLockPath, "{}"); err != nil {
-			log.Warningf("Failed to unlock lockFirst shard after failed lockSecond lock for %v/%v", keyspace, shard)
+		if err := fLD.Unlock(ctx); err != nil {
+			log.Warningf("Failed to unlock lockFirst after failed lockSecond lock for %v %v: %v", cell, dirPath, err)
```
Comment (pivanof) [inline go/vt/topo/helpers/tee.go]: Still no comma here.
